tests/scripts/all.sh

```diff
@@ -2057,6 +2059,7 @@ component_test_psa_crypto_config_accel_rsa_signature () {
     scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_RIPEMD160_C
 
     scripts/config.py -f tests/include/test/drivers/config_test_driver.h set MBEDTLS_SHA1_C
+    scripts/config.py -f tests/include/test/drivers/config_test_driver.h set MBEDTLS_SHA224_C
```

Comment:
Is this needed? What was failing?

Reply:
Since waiting for the CI takes a lot of time, I run some local test before creating the commit with teh following command:
`./tests/scripts/all.sh --force test_psa_crypto_config_accel_*`

From there I manually set these changes to all the tests that were failing until all of them were happy. In detail I don't remember which one was failing here, but I can give it a try to let you know

Reply:
It would be great to know what's the exact problem here - maybe you discovered some misaligned dependencies that have to be fixed? Maybe there are some tests that depend on SHA256 while they should depend on SHA224? That's why I'm digging :)

Reply:
if I remove this specific change and then I run the
`./tests/scripts/all.sh --force test_psa_crypto_config_accel_rsa_signature`
test, then I get the following errors:
```
read type: RSA_PAIR 1024-bit, RSA_PSS(SHA_224) .................... FAILED
read type: RSA_PAIR 1024-bit, RSA_PSS_ANY_SALT(SHA_224) ........... FAILED
read type: RSA_PAIR 1536-bit, RSA_PSS(SHA_224) .................... FAILED
read type: RSA_PAIR 1536-bit, RSA_PSS_ANY_SALT(SHA_224) ........... FAILED
```

Reply:
There's probably some kind of an underlying issue here: if these tests are run, then `PSA_WANT_ALG_SHA_224` is on (that's the proper dependency), but if they fail, the builtin /driver implementation isn't provided. I'm not convinced that simply adding `MBEDTLS_SHA224_C` is the solution here - I'll look further into it.

Reply:
sorry but I'm bit confused now: isn't it the same also for `SHA1_C` and `SHA512_C´? 
They should be included as `PSA_WANT_ALG_SHA_*` in `crypto_config.h`, but they are manually set before building the test library (even though I don't understand why, because your observation makes sense to me)

Reply:
I don't think there's an issue here. 

> maybe you discovered some misaligned dependencies that have to be fixed? Maybe there are some tests that depend on SHA256 while they should depend on SHA224?

I don't think so. The need to explicitly enable SHA224 here appeared when Valerio removed it from the base config in `config_test_driver.h`, not when we made it independent from SHA-256. It makes sense that if it's disabled by default, we need to enable it when we need it (like the other hashes in this test, as Valerio points out).

> if these tests are run, then `PSA_WANT_ALG_SHA_224` is on (that's the proper dependency), but if they fail, the builtin / driver implementation isn't provided.

It's slightly more complex due to the fact the RSA-PSS is a composite algorithm, as show in the PSA macros for it (and the test descriptions above): it's actually RSA-PSS-SHA256, or RSA-PSS-SHA224, etc. And it's not just that RSA-PSS-SHA224 expects the output of SHA-224 as its input, it's that it performs additional SHA-224 computations as part PSS padding. So, when we accelerate RSA-PSS signatures (what this component is doing), the RSA-PSS driver needs to be able to compute SHA224 if you're going to ask it to perform RSA-PSS-SHA224 signatures.

That's why SHA-224 (and again, all the other hashes long before this PR), need to be enabled in libtestdriver1. Now, that's done using `MBEDTLS_SHA224_C` because some dark magic happens when building libtestdriver1 and `PSA_WANT` macros don't work the way they usually do there, see `tests/include/test/drivers/crypto_config_test_driver_extension.h`, but that magic doesn't affect legacy macros. (Also, RSA-PSS uses the legacy API for doing its hashing, mostly.)

So, basically, `PSA_WANT_ALG_SHA_224` in the *main libraries* build, means you're going to run not skip the RSA-PSS-SHA224 tests, which means you need the SHA-224 implementation in libtestdriver1 that provides RSA-PSS, which means we need `MBEDTLS_SHA224_C` enabled there, and just having `PSA_WANT_ALG_SHA_224` set when building libtestdriver1 won't achieve that (as it would when building the main libraries).

> That's why I'm digging :)

Please keep digging if you're not fully convinced by my explanations. However, if your digging uncovers some issue, I suggest we treat it as pre-existing and not block this PR on it (after all, it's only doing for SHA-224 what the component was already doing from SHA-384).

Reply:
Thanks a lot for the detailed explanation! Now the difference between _main library_'s build and _libtestdriver1_'s one is more clear, but it's true that there's some dark magic in between that confused me yesterday during the analysis :D

Reply:
Actually I don't understand all the consequences of the magic happening when building libtestdriver1, and I have questions about that for Ronald when he comes back. Once I understand better, I'll try to write some documentation about it (unless Ronald does), because the current level of documentation seems a bit low compared to the complexity/subtlety of the thing.

Reply:
Thanks, @mpg! This part is pretty clear to me, what isn't clear is what happens in `crypto_config_test_driver_extension.h`, which turns off `MBEDTLS_PSA_ACCEL_ALG_SHA_224` (or other algs, respectively) if it's already defined, and adds it otherwise, if `PSA_WANT_ALG_SHA_224` is on. 
As I understand it: there's probably no partial capability definition for a driver, and the config used for the library & apps build later on matters (hence it has `PSA_WANT_ALG_SHA_224`).

Reply:
Here's my current understanding of what happens in `crypto_config_test_driver_extension.h`: the things we want accelerated in the main libraries are the things we want to be built-in in the driver library; conversely, things that we don't want accelerated in the main libraries don't need to be present as built-in in the driver library. However, according to https://github.com/Mbed-TLS/mbedtls/pull/5089 (second to last paragraph) we need the `PSA_WANT` symbols to be the same in the driver library and the main library, so for the things we don't want to have in the driver library, instead of disabling the corresponding `PSA_WANT` macro, we set the corresponding `MBEDTLS_PSA_ACCEL` macro, which disables the corresponding `MBEDTLS_PSA_BUILTIN` macro.

So, the when you look at the `MBEDTLS_PSA_ACCEL` macro, their state is opposite between the driver library and the main library. This reversal is what `crypto_config_test_driver_extension.h` performs (the comment at the top of the file says "mirrors" which I think is not the best choice of words to express that).

However my understanding is quite incomplete at this point. For example, it's not clear to me why the `PSA_WANT` macros need to be the same in driver and main; it's even less clear to me why we need to unset `PSA_WANT_ALG_STREAM`  and `ECB_NO_PADDING` everywhere... As I said, I'm gonna have a chat with Ronald and then ensure some documentation gets written about it.

Reply:
> I don't understand all the consequences of the magic happening when building libtestdriver1

Neither do I. I do know that some of it is to missing `#if defined` or `depends_on` in various places that have since then been fixed.

> because the current level of documentation seems a bit low compared to the complexity/subtlety of the thing.

> why we need to unset PSA_WANT_ALG_STREAM and ECB_NO_PADDING everywhere

I think we can simplify the build. In particular, the configuration changes (other than `ACCEL` on/off of course) are hacks that hopefully aren't necessary anymore, or that are now close to being fixable (thanks to more driver work, `depends.py`, this very SHA cleanup, etc.). That should make it possible to just run `make SOMETHING_SIMPLE` rather than the complex script in `all.sh`.

> why the PSA_WANT macros need to be the same in driver and main

For one thing, buffer sizes in `crypto_sizes.h` and context structures in `crypto_struct.h` are part of the shared ABI between driver and main.